mscore/seq.cpp

```diff
@@ -1405,10 +1415,13 @@ void Seq::putEvent(const NPlayEvent& event, unsigned framePos)
             qDebug("bad channel value %d >= %d", channel, cs->midiMapping()->size());
             return;
             }
+
+      // dispatch midi messages to MIDI driver first before rendering with MuseScore's synth, in order to avoid additional unpredictable latency associated with MuseScore's synth
+      if (_driver != 0 && (preferences.useJackMidi || preferences.useAlsaAudio || preferences.usePortaudioAudio))
+            _driver->putEvent(event, framePos);
```

Comment:
I'm a little suspicious of Fluid::alloc_voice() which does the following:

      Voice* v = freeVoices.takeLast();
      activeVoices.append(v);

Note that the first actually operation performs a delete, and the second operation performs an allocation, but this is a problem because as the header above Seq::process says:

```
//    This function is called in a realtime context. This
//    means that no blocking operations are allowed which
//    includes memory allocation. The usual thread synchronisation
//    methods like semaphores can also not be used.
```

Now clearly looking at the operation of that alloc_voices, it seems that should utilize what is known as a "Ring Buffer".  I'm looking for a generic qt class for this...all I can find is: https://doc.qt.io/archives/qt-5.5/qt3d-qcircularbuffer.html but for some reason that includes a dependency on qt3d (which I don't know why Qt did that) and I don't think should include Qt3D.  I might just use a simple C-style pre-allocated array and have a first index and last index...I'll just have to make sure the array is large enough to hold all voices that need to be used at the same time.